File: d1c_create_graph_conn.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <sstream>
#include <regex>
#include "C:/clib/mysql/include/mysql.h"
#include "nlohmann/json.hpp"

using json = nlohmann::json;

class GraphConnCreator {
private:
    MYSQL* conn;
    std::string mysql_host;
    std::string mysql_user;
    std::string mysql_password;
    std::string mysql_database;
    std::string neo4j_uri;
    std::string neo4j_username;
    std::string neo4j_password;
    std::string csv_file;

    std::string convertHttpToBoltUri(const std::string& http_uri) {
        std::regex http_pattern("http://(.*?)(:\\d+)?(/.*)?");
        std::smatch matches;
        
        if (std::regex_match(http_uri, matches, http_pattern)) {
            std::string host = matches[1].str();
            std::string port = matches[2].str();
            
            if (port.empty()) {
                port = ":7687";
            } else if (port == ":7474") {
                port = ":7687";
            }
            
            return "bolt://" + host + port;
        }
        
        return "bolt://" + http_uri;
    }

    bool loadConfig() {
        try {
            std::ifstream configFile("config.json");
            if (!configFile.is_open()) {
                std::cerr << "无法打开config.json文件" << std::endl;
                return false;
            }

            json config;
            configFile >> config;

            mysql_host = config["mysql"]["host"];
            mysql_user = config["mysql"]["user"];
            mysql_password = config["mysql"]["password"];
            mysql_database = config["mysql"]["database"];

            neo4j_uri = config["neo4j"]["uri"];
            neo4j_username = config["neo4j"]["username"];
            neo4j_password = config["neo4j"]["password"];
            
            if (neo4j_uri.find("http://") == 0) {
                std::string bolt_uri = convertHttpToBoltUri(neo4j_uri);
                std::cout << "将Neo4j HTTP URI转换为Bolt URI: " << neo4j_uri << " -> " << bolt_uri << std::endl;
                neo4j_uri = bolt_uri;
            } else if (neo4j_uri.find("bolt://") != 0 && neo4j_uri.find("neo4j://") != 0) {
                neo4j_uri = "bolt://" + neo4j_uri;
                std::cout << "添加bolt://前缀到Neo4j URI: " << neo4j_uri << std::endl;
            }

            return true;
        }
        catch (const std::exception& e) {
            std::cerr << "读取配置文件失败: " << e.what() << std::endl;
            return false;
        }
    }

    bool connectToMysql() {
        conn = mysql_init(NULL);
        if (conn == NULL) {
            std::cerr << "MySQL初始化失败" << std::endl;
            return false;
        }

        if (!mysql_real_connect(conn, mysql_host.c_str(), mysql_user.c_str(), 
            mysql_password.c_str(), mysql_database.c_str(), 0, NULL, 0)) {
            std::cerr << "MySQL连接失败: " << mysql_error(conn) << std::endl;
            return false;
        }

        mysql_set_character_set(conn, "utf8");
        return true;
    }

    bool exportConnectionsToCSV() {
        csv_file = "output/connections_export.csv";
        
        system("if not exist output mkdir output");
        
        std::ofstream csvOutput(csv_file);
        if (!csvOutput.is_open()) {
            std::cerr << "无法创建CSV文件: " << csv_file << std::endl;
            return false;
        }
        
        csvOutput << "source,target,connNo,connType,color,isCable,voltage,current,resistance" << std::endl;
        
        if (mysql_query(conn, "SELECT source, target, connNo, connType, color, isCable, voltage, current, resistance FROM conn_graph WHERE isInPanel=1")) {
            std::cerr << "查询conn_graph失败: " << mysql_error(conn) << std::endl;
            return false;
        }

        MYSQL_RES* result = mysql_store_result(conn);
        if (result == NULL) {
            std::cerr << "获取结果集失败" << std::endl;
            return false;
        }

        int num_fields = mysql_num_fields(result);
        MYSQL_ROW row;
        int count = 0;
        
        while ((row = mysql_fetch_row(result))) {
            for (int i = 0; i < num_fields; i++) {
                if (i > 0) csvOutput << ",";
                if (row[i]) {
                    std::string value = row[i];
                    if (value.find(',') != std::string::npos || 
                        value.find('"') != std::string::npos) {
                        size_t pos = 0;
                        while ((pos = value.find('"', pos)) != std::string::npos) {
                            value.replace(pos, 1, "\"\"");
                            pos += 2;
                        }
                        csvOutput << "\"" << value << "\"";
                    } else {
                        csvOutput << value;
                    }
                } else {
                    csvOutput << "";
                }
            }
            csvOutput << std::endl;
            count++;
        }

        mysql_free_result(result);
        csvOutput.close();
        
        std::cout << "已导出 " << count << " 条连接记录到CSV文件: " << csv_file << std::endl;
        return true;
    }

    bool createConnectionsInNeo4j() {
        std::string pythonScript = "import_connections_to_neo4j.py";
        std::ofstream scriptFile(pythonScript);
        
        if (!scriptFile.is_open()) {
            std::cerr << "无法创建Python脚本文件" << std::endl;
            return false;
        }
        
        scriptFile << "import csv\n";
        scriptFile << "import sys\n";
        scriptFile << "from neo4j import GraphDatabase\n\n";
        
        scriptFile << "uri = '" << neo4j_uri << "'\n";
        scriptFile << "username = '" << neo4j_username << "'\n";
        scriptFile << "password = '" << neo4j_password << "'\n\n";
        
        scriptFile << "print(f'连接到Neo4j数据库: {uri}')\n";
        scriptFile << "try:\n";
        scriptFile << "    driver = GraphDatabase.driver(uri, auth=(username, password))\n";
        scriptFile << "    \n";
        scriptFile << "    with driver.session() as session:\n";
        scriptFile << "        result = session.run('RETURN 1 AS test')\n";
        scriptFile << "        test_value = result.single()['test']\n";
        scriptFile << "        print(f'连接测试成功: {test_value}')\n\n";
        
        scriptFile << "        # 检查节点数量\n";
        scriptFile << "        result = session.run('MATCH (d:V_Device) RETURN count(d) AS deviceCount')\n";
        scriptFile << "        device_count = result.single()['deviceCount']\n";
        scriptFile << "        print(f'数据库中存在 {device_count} 个V_Device节点')\n\n";
        
        scriptFile << "        result = session.run('MATCH (t:V_terminal) RETURN count(t) AS terminalCount')\n";
        scriptFile << "        terminal_count = result.single()['terminalCount']\n";
        scriptFile << "        print(f'数据库中存在 {terminal_count} 个V_Terminal节点')\n\n";
        
        scriptFile << "        # 清空现有连接关系\n";
        scriptFile << "        result = session.run('MATCH ()-[r:CONN]->() DELETE r')\n";
        scriptFile << "        print('已清除所有现有连接关系')\n\n";
        
        scriptFile << "    # 从CSV文件导入连接关系\n";
        scriptFile << "    conn_count = 0\n";
        scriptFile << "    fail_count = 0\n";
        scriptFile << "    with open('" << csv_file << "', 'r', encoding='utf-8') as file:\n";
        scriptFile << "        reader = csv.DictReader(file)\n";
        scriptFile << "        for row in reader:\n";
        scriptFile << "            with driver.session() as session:\n";
        scriptFile << "                cypher = '''\n";
        scriptFile << "                    MATCH (source)\n";
        scriptFile << "                    WHERE source.fdid = $source\n";
        scriptFile << "                    AND (source:V_Device OR source:V_terminal)\n";
        scriptFile << "                    MATCH (target)\n";
        scriptFile << "                    WHERE target.fdid = $target\n";
        scriptFile << "                    AND (target:V_Device OR target:V_terminal)\n";
        scriptFile << "                    CREATE (source)-[r:CONN {\n";
        scriptFile << "                        connNo: $connNo,\n";
        scriptFile << "                        type: $connType,\n";
        scriptFile << "                        color: $color,\n";
        scriptFile << "                        isCable: $isCable,\n";
        scriptFile << "                        voltage: $voltage,\n";
        scriptFile << "                        current: $current,\n";
        scriptFile << "                        resistance: $resistance\n";
        scriptFile << "                    }]->(target)\n";
        scriptFile << "                    RETURN r\n";
        scriptFile << "                '''\n";
        scriptFile << "                try:\n";
        scriptFile << "                    result = session.run(\n";
        scriptFile << "                        cypher,\n";
        scriptFile << "                        source=row['source'],\n";
        scriptFile << "                        target=row['target'],\n";
        scriptFile << "                        connNo=row['connNo'],\n";
        scriptFile << "                        connType=row['connType'],\n";
        scriptFile << "                        color=row['color'],\n";
        scriptFile << "                        isCable=row['isCable'] == '1',\n";
        scriptFile << "                        voltage=float(row['voltage'] or 0),\n";
        scriptFile << "                        current=float(row['current'] or 0),\n";
        scriptFile << "                        resistance=float(row['resistance'] or 0)\n";
        scriptFile << "                    )\n";
        scriptFile << "                    if not result.peek():\n";
        scriptFile << "                        print('未能找到节点: source=' + row['source'] + ', target=' + row['target'])\n";
        scriptFile << "                        fail_count += 1\n";
        scriptFile << "                    else:\n";
        scriptFile << "                        conn_count += 1\n";
        scriptFile << "                        if conn_count % 100 == 0:\n";
        scriptFile << "                            print(f'已创建 {conn_count} 个连接关系')\n";
        scriptFile << "                except Exception as e:\n";
        scriptFile << "                    print(f'创建连接失败: {e}', file=sys.stderr)\n";
        scriptFile << "                    fail_count += 1\n\n";
        
        scriptFile << "    print(f'总共创建了 {conn_count} 个连接关系，失败 {fail_count} 个')\n\n";
        scriptFile << "    # 验证连接关系数量\n";
        scriptFile << "    with driver.session() as session:\n";
        scriptFile << "        result = session.run('MATCH ()-[r:CONN]->() RETURN count(r) AS connCount')\n";
        scriptFile << "        conn_count = result.single()['connCount']\n";
        scriptFile << "        print(f'数据库中实际存在 {conn_count} 个CONN关系')\n\n";
        
        scriptFile << "    driver.close()\n";
        scriptFile << "except Exception as e:\n";
        scriptFile << "    print(f'错误: {e}', file=sys.stderr)\n";
        scriptFile << "    sys.exit(1)\n";
        
        scriptFile.close();
        
        std::cout << "执行Python脚本导入数据到Neo4j..." << std::endl;
        int result = system("python --version > nul 2>&1");
        
        if (result != 0) {
            std::cerr << "未找到Python命令，尝试使用python3命令..." << std::endl;
            result = system("python3 --version > nul 2>&1");
            
            if (result != 0) {
                std::cerr << "未找到Python3命令，请确保Python已安装并添加到PATH环境变量中" << std::endl;
                return false;
            }
            
            result = system("python3 -m pip install neo4j > nul 2>&1");
            if (result != 0) {
                std::cerr << "无法安装Neo4j Python驱动，请手动运行: pip install neo4j" << std::endl;
            }
            
            result = system(("python3 " + pythonScript).c_str());
        } else {
            result = system("python -m pip install neo4j > nul 2>&1");
            if (result != 0) {
                std::cerr << "无法安装Neo4j Python驱动，请手动运行: pip install neo4j" << std::endl;
            }
            
            result = system(("python " + pythonScript).c_str());
        }
        
        if (result != 0) {
            std::cerr << "执行Python脚本失败，错误代码: " << result << std::endl;
            std::cerr << "请确保已安装Neo4j Python驱动: pip install neo4j" << std::endl;
            return false;
        }
        
        return true;
    }

    bool createGraphFromConnections() {
        if (!exportConnectionsToCSV()) {
            return false;
        }
        
        if (!createConnectionsInNeo4j()) {
            return false;
        }
        
        return true;
    }

public:
    GraphConnCreator() : conn(nullptr) {}
    
    ~GraphConnCreator() {
        if (conn) {
            mysql_close(conn);
        }
    }

    bool run() {
        if (!loadConfig()) {
            return false;
        }

        if (!connectToMysql()) {
            return false;
        }

        bool success = createGraphFromConnections();

        return success;
    }
};

int main() {
    GraphConnCreator creator;
    if (!creator.run()) {
        std::cerr << "程序执行失败" << std::endl;
        return 1;
    }
    
    std::cout << "成功创建连接关系" << std::endl;
    return 0;
}

File: d1c_create_graph_conn.py
import pymysql
import sys
import os
import traceback
from neo4j import GraphDatabase
from config import MYSQL_CONFIG, NEO4J_URI, NEO4J_USER, NEO4J_PASSWORD

# 添加 MySQL 连接超时和字符集配置
MYSQL_CONFIG.update({
    "connect_timeout": 10,
    "charset": 'utf8mb4'
})

def log_message(message, file_handle=None):
    """打印并记录日志消息"""
    print(message, flush=True)
    if file_handle:
        file_handle.write(message + '\n')
        file_handle.flush()

def main():
    # 创建日志目录
    if not os.path.exists('logs'):
        os.makedirs('logs')

    log_file = 'logs/conn_creation.log'
    with open(log_file, 'w', encoding='utf-8') as f:
        try:
            log_message(f'连接到Neo4j数据库: {NEO4J_URI}', f)
            
            # 连接MySQL数据库
            log_message('尝试连接MySQL数据库...', f)
            try:
                mysql_conn = pymysql.connect(**MYSQL_CONFIG, cursorclass=pymysql.cursors.DictCursor)
                log_message('MySQL连接成功', f)
            except pymysql.Error as e:
                log_message(f'MySQL连接错误: {str(e)}', f)
                if hasattr(e, 'errno'):
                    log_message(f'错误代码: {e.errno}', f)
                raise

            # 连接Neo4j数据库
            driver = GraphDatabase.driver(NEO4J_URI, auth=(NEO4J_USER, NEO4J_PASSWORD))
            
            # 测试Neo4j连接
            with driver.session() as session:
                result = session.run('RETURN 1 AS test')
                test_value = result.single()['test']
                log_message(f'Neo4j连接测试成功: {test_value}', f)

                # 清除现有连接关系
                result = session.run('MATCH ()-[r:CONN]->() DELETE r')
                log_message('已清除所有现有连接关系', f)

                # 检查Neo4j中的节点数量
                result = session.run("""
                    MATCH (n:V_Terminal)
                    RETURN count(n) as count
                """)
                node_count = result.single()['count']
                log_message(f"\nNeo4j中存在 {node_count} 个V_Terminal节点", f)

                # 检查几个关键节点的存在性
                log_message("\n检查示例节点:", f)
                check_nodes = [
                    "=A01+K1.H2-Q1:D2",
                    "=A01+K1.B1-K3:14",
                    "=Q15+K1.G1-G5:ANALOG.15",
                    "=S02+K1.H2-Q1:2"
                ]
                for ftid in check_nodes:
                    result = session.run("""
                        MATCH (n:V_Terminal {FTID: $ftid})
                        RETURN n.FTID as ftid, labels(n) as labels, properties(n) as props
                    """, ftid=ftid)
                    if result.peek():
                        record = result.single()
                        log_message(f"\n找到节点: {ftid}", f)
                        log_message(f"标签: {record['labels']}", f)
                        log_message(f"属性: {record['props']}", f)
                    else:
                        log_message(f"\n未找到节点: {ftid}", f)

            # 从MySQL读取连接数据
            cursor = mysql_conn.cursor()
            
            # 获取总记录数
            cursor.execute("SELECT COUNT(*) as count FROM conn_graph")
            total_count = cursor.fetchone()['count']
            log_message(f"\n总共需要处理 {total_count} 条连接记录", f)
            
            # 分批处理所有记录
            batch_size = 1000
            processed = 0
            conn_count = 0
            fail_count = 0
            
            while processed < total_count:
                cursor.execute(f"""
                    SELECT * FROM conn_graph 
                    LIMIT {processed}, {batch_size}
                """)
                
                rows = cursor.fetchall()
                if not rows:
                    break
                
                for row in rows:
                    source = row['source']
                    target = row['target']
                    
                    # 在Neo4j中创建连接
                    with driver.session() as session:
                        # 先检查节点
                        source_result = session.run("""
                            MATCH (n:V_Terminal {FTID: $ftid})
                            RETURN n
                        """, ftid=source)
                        target_result = session.run("""
                            MATCH (n:V_Terminal {FTID: $ftid})
                            RETURN n
                        """, ftid=target)

                        source_exists = source_result.peek() is not None
                        target_exists = target_result.peek() is not None

                        if not source_exists or not target_exists:
                            fail_count += 1
                            log_message(f"\n节点缺失: source={source} ({source_exists}), target={target} ({target_exists})", f)
                            # 获取节点的详细信息
                            if not source_exists:
                                result = session.run("""
                                    MATCH (n:V_Terminal)
                                    WHERE n.FTID STARTS WITH $prefix
                                    RETURN n.FTID LIMIT 5
                                """, prefix=source.split(":")[0])
                                similar = [r['n.FTID'] for r in result]
                                if similar:
                                    log_message(f"类似的source节点: {similar}", f)
                            if not target_exists:
                                result = session.run("""
                                    MATCH (n:V_Terminal)
                                    WHERE n.FTID STARTS WITH $prefix
                                    RETURN n.FTID LIMIT 5
                                """, prefix=target.split(":")[0])
                                similar = [r['n.FTID'] for r in result]
                                if similar:
                                    log_message(f"类似的target节点: {similar}", f)
                            continue

                        # 准备连接属性
                        props = {k: v for k, v in row.items() if k not in ['source', 'target']}
                        for k in ['voltage', 'current', 'resistance']:
                            props[k] = float(props.get(k, 0))
                        props['isCable'] = bool(props.get('isCable', 0))
                        props['isInPanel'] = bool(props.get('isInPanel', 0))

                        # 创建连接
                        props_str = ', '.join(f'{k}: ${k}' for k in props.keys())
                        cypher = f"""
                            MATCH (source:V_Terminal {{FTID: $source}})
                            MATCH (target:V_Terminal {{FTID: $target}})
                            CREATE (source)-[r:CONN {{{props_str}}}]->(target)
                            RETURN r
                        """
                            
                        try:
                            result = session.run(cypher, dict(source=source, target=target, **props))
                            if result.peek():
                                conn_count += 1
                            else:
                                fail_count += 1
                                log_message(f"\n创建连接失败: {source} -> {target}", f)
                        except Exception as e:
                            fail_count += 1
                            log_message(f"\n创建连接时出错: {source} -> {target}", f)
                            log_message(f"错误信息: {e}", f)
                    
                    processed += 1
                    if processed % 100 == 0:
                        progress = (processed / total_count) * 100
                        log_message(f"\r处理进度: {processed}/{total_count} ({progress:.1f}%) - 成功: {conn_count}, 失败: {fail_count}", f)

            log_message(f'\n总共创建了 {conn_count} 个连接关系，失败 {fail_count} 个', f)

            # 验证最终的连接数量
            with driver.session() as session:
                result = session.run('MATCH ()-[r:CONN]->() RETURN count(r) AS count')
                final_count = result.single()['count']
                log_message(f'\n数据库中实际存在 {final_count} 个CONN关系', f)

            # 关闭连接
            mysql_conn.close()
            driver.close()
            log_message('\n程序执行成功', f)
            return 0

        except Exception as e:
            log_message(f'错误: {str(e)}', f)
            log_message('详细错误信息:', f)
            log_message(traceback.format_exc(), f)
            return 1

if __name__ == "__main__":
    sys.exit(main())